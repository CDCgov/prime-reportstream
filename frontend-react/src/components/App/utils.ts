/**
 * Type-safe way to ensure refs are initialized only once.
 */
export function getService<T>(
    ref: React.MutableRefObject<T>,
    initFn: () => Exclude<T, null>,
) {
    if (ref.current !== null) {
        return ref.current;
    }
    const service = initFn();
    ref.current = service;
    return service;
}
