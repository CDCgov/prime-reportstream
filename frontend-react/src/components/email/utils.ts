export function oktaProperty(name: string) {
    return `$\{${name}}`;
}
