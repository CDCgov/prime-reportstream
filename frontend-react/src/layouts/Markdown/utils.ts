import { mergeWith } from "lodash";
import { AppConfig } from "../../config";

/**
 * Deep merge default meta with frontmatter meta. Has customizer that takes empty strings into account.
 */
export function createMeta(config: AppConfig, frontmatter: Frontmatter) {
    return mergeWith(
        {},
        config.PAGE_META.defaults,
        frontmatter.meta,
        (a, b) => {
            if (typeof b === "string" && b === "") return a;
            return undefined;
        },
    );
}
