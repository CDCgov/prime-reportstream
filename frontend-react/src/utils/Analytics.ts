import { getAppInsights } from "../TelemetryService";

export enum EventName {
    TABLE_FILTER = "Table Filter",
    SESSION_DURATION = "Session Duration",
    TABLE_PAGINATION = "Table Pagination",
}

export const trackAppInsightEvent = (eventName: string, eventData: any) => {
    const appInsights = getAppInsights();
    const telemetryEvent = {
        name: eventName,
        properties: eventData,
    };

    if (eventName !== "") appInsights?.trackEvent(telemetryEvent);
};
