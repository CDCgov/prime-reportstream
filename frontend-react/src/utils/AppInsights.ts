export enum EventName {
    TABLE_FILTER = "Table Filter",
    SESSION_DURATION = "Session Duration",
    TABLE_PAGINATION = "Table Pagination",
    FILE_VALIDATOR = "File Validator",
}
