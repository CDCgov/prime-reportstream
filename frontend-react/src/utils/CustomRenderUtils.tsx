import { ReactElement, ReactNode, Suspense } from "react";
import {
    render,
    RenderOptions,
    renderHook as renderHookOrig,
    RenderHookOptions,
    Queries,
    queries,
} from "@testing-library/react";
import {
    createMemoryRouter,
    Outlet,
    RouterProvider,
    RouteObject,
} from "react-router-dom";
import { QueryClientProvider } from "@tanstack/react-query";
import { HelmetProvider } from "react-helmet-async";
import { Fixture, MockResolver } from "@rest-hooks/test";
import { CacheProvider } from "rest-hooks";

import { SessionProviderBase } from "../contexts/Session";
import { AuthorizedFetchProvider } from "../contexts/AuthorizedFetch";
import { getTestQueryClient } from "../network/QueryClients";
import { FeatureFlagProvider } from "../contexts/FeatureFlag";
import { appRoutes } from "../AppRouter";
import AppInsightsContextProvider from "../contexts/AppInsights";
import config from "../config";

interface AppWrapperProps {
    children: ReactNode;
}

interface AppWrapperOptions {
    initialRouteEntries?: string[];
    restHookFixtures?: Fixture[];
}

function TestLayout() {
    return <Outlet />;
}

function createTestRoutes(
    routes: RouteObject[],
    element: ReactNode,
): RouteObject[] {
    return routes.map((r) => ({
        ...r,
        lazy: undefined,
        element: r.path !== "/" ? element : <TestLayout />,
        children: r.children
            ? createTestRoutes(r.children, element)
            : undefined,
    })) as RouteObject[];
}
export const AppWrapper = ({
    initialRouteEntries,
    restHookFixtures,
}: AppWrapperOptions = {}) => {
    // FUTURE_TODO: Replace children with <AppRouter /> if initialRouteEntries after mocking okta users
    // in tests is made easier for better coverage as we'd be able to test through
    // any custom route wrappers.
    // FUTURE_TODO: Remove MockResolver and restHookFixtures when removing react-hooks.
    return ({ children }: AppWrapperProps) => {
        /**
         * Dynamically makes the supplied children the return element for all
         * routes.
         * FUTURE_TODO: Remove this once okta user/session mocking is easier
         * and use <AppRouter /> instead.
         */
        const router = createMemoryRouter(
            createTestRoutes(appRoutes, children),
            {
                initialEntries: initialRouteEntries,
            },
        );
        return (
            <Suspense>
                <CacheProvider>
                    <HelmetProvider>
                        <AppInsightsContextProvider>
                            <SessionProviderBase
                                oktaAuth={{} as any}
                                authState={{}}
                                config={config}
                            >
                                <QueryClientProvider
                                    client={getTestQueryClient()}
                                >
                                    <AuthorizedFetchProvider
                                        initializedOverride={true}
                                    >
                                        <FeatureFlagProvider>
                                            {restHookFixtures ? (
                                                <MockResolver
                                                    fixtures={restHookFixtures}
                                                >
                                                    <RouterProvider
                                                        router={router}
                                                    />
                                                </MockResolver>
                                            ) : (
                                                <RouterProvider
                                                    router={router}
                                                />
                                            )}
                                        </FeatureFlagProvider>
                                    </AuthorizedFetchProvider>
                                </QueryClientProvider>
                            </SessionProviderBase>
                        </AppInsightsContextProvider>
                    </HelmetProvider>
                </CacheProvider>
            </Suspense>
        );
    };
};

interface RenderAppOptions extends RenderOptions, AppWrapperOptions {}

export const renderApp = (
    ui: ReactElement,
    {
        initialRouteEntries,
        restHookFixtures,
        ...options
    }: Omit<RenderAppOptions, "wrapper"> = {},
) => {
    return render(ui, {
        wrapper: AppWrapper({ initialRouteEntries, restHookFixtures }),
        ...options,
    });
};

export function renderHook<
    Result,
    Props,
    Q extends Queries = typeof queries,
    Container extends Element | DocumentFragment = HTMLElement,
    BaseElement extends Element | DocumentFragment = Container,
>(
    render: (initialProps: Props) => Result,
    options?: RenderHookOptions<Props, Q, Container, BaseElement>,
) {
    return renderHookOrig<Result, Props, Q, Container, BaseElement>(render, {
        wrapper: AppWrapper(),
        ...options,
    });
}

export { screen } from "@testing-library/react";
