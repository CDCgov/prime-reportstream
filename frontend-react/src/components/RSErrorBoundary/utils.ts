import { ErrorBoundaryProps } from "react-error-boundary";
import { RSConsole } from "../../utils/rsConsole/rsConsole";
import { isRSNetworkError } from "../../utils/RSNetworkError";

/**
 * Error handler that emits through rsConsole if available.
 */
export function createErrorLogger(
    rsConsole?: RSConsole,
): Exclude<ErrorBoundaryProps["onError"], undefined> {
    return (exception, info) => {
        if (!isRSNetworkError(exception)) {
            console.warn(
                "Please work to migrate all non RSError throws to use an RSError object.",
            );
        }
        if (rsConsole) {
            // React will always console.error all errors, regardless of boundary,
            // so just emit the telemetry.
            rsConsole._error(
                {
                    args: [exception, info.componentStack],
                    location: window.location.href,
                },
                rsConsole.severityLevels.error,
            );
        }
    };
}
