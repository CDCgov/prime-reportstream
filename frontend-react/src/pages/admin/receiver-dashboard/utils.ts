import { addMinutes, eachDayOfInterval, interval } from "date-fns";
import { RSReceiverStatus } from "../../../hooks/api/UseReceiversConnectionStatus/UseReceiversConnectionStatus";
import { DatePair } from "../../../utils/DateTimeUtils";

export enum SuccessRate {
    UNDEFINED = "UNDEFINED",
    ALL_SUCCESSFUL = "ALL_SUCCESSFUL",
    ALL_FAILURE = "ALL_FAILURE",
    MIXED_SUCCESS = "MIXED_SUCCESS",
}

export const SUCCESS_RATE_CLASSNAME_MAP = {
    [SuccessRate.UNDEFINED]: "success-undefined",
    [SuccessRate.ALL_SUCCESSFUL]: "success-all",
    [SuccessRate.ALL_FAILURE]: "failure-all",
    [SuccessRate.MIXED_SUCCESS]: "success-mixed",
};

export enum MatchingFilter {
    NO_FILTER,
    FILTER_NOT_MATCHED,
    FILTER_IS_MATCHED,
}

export const MATCHING_FILTER_CLASSNAME_MAP = {
    [MatchingFilter.NO_FILTER]: "",
    [MatchingFilter.FILTER_NOT_MATCHED]: "success-result-hidden",
    [MatchingFilter.FILTER_IS_MATCHED]: "",
};

/**
 * build the dictionary with a special path+key
 * @param dataIn
 */
export const sortStatusData = <
    T extends RSReceiverStatus[] | RSReceiverStatusParsed[],
>(
    dataIn: T,
): T => {
    const data = structuredClone(dataIn);
    const { orgNameMaxLength, receiverNameMaxLength } = data.reduce(
        (prev, curr) => {
            prev.orgNameMaxLength =
                curr.organizationName.length > prev.orgNameMaxLength
                    ? curr.organizationName.length
                    : prev.orgNameMaxLength;
            prev.receiverNameMaxLength =
                curr.receiverName.length > prev.receiverNameMaxLength
                    ? curr.receiverName.length
                    : prev.receiverNameMaxLength;

            return prev;
        },
        {
            orgNameMaxLength: 0,
            receiverNameMaxLength: 0,
        },
    );

    // sorting by organizationName, then receiverName, then connectionCheckStartedAt
    data.sort(
        (
            d1: RSReceiverStatus | RSReceiverStatusParsed,
            d2: RSReceiverStatus | RSReceiverStatusParsed,
        ) => {
            // ideally the shape of this type will change so that all receivers, regardless if any status data was found, will get sorted
            const [sortStrA, sortStrB] = [d1, d2].map(
                (x) =>
                    `${x.organizationName.padEnd(orgNameMaxLength, "-")}${x.receiverName.padEnd(receiverNameMaxLength, "-")}${new Date(x.connectionCheckStartedAt || Date.now()).toISOString()}`,
            );

            const result =
                sortStrA > sortStrB ? 1 : sortStrA < sortStrB ? -1 : 0;

            return result;
        },
    );

    return data;
};

export interface TimePeriodData {
    start: Date;
    end: Date;
    success: number;
    fail: number;
    isResultFilterMatch: boolean;
    successRateType: SuccessRate;
    matchingFilter: MatchingFilter;
    entries: RSReceiverStatusParsed[];
}

export function createStatusTimePeriodData({
    data,
    range: [startDate, endDate],
    filterResultMessage = "",
    timePeriodMinutes = 2 * 60,
}: {
    data: RSReceiverStatus[];
    range: DatePair;
    filterResultMessage?: string;
    timePeriodMinutes?: number;
}): ReceiverStatusTimePeriod[] {
    const inter = interval(startDate, endDate, { assertPositive: true });
    const numTimePeriodsPerDay = (24 * 60) / timePeriodMinutes;
    if (numTimePeriodsPerDay % 1 !== 0)
        throw new Error("Invalid time period duration");

    const timePeriodLabels = Array.from(Array(numTimePeriodsPerDay).keys()).map(
        (_, i) =>
            addMinutes(
                startDate,
                (i + 1) * timePeriodMinutes,
            ).toLocaleTimeString(),
    );
    const sortedData = sortStatusData(data).map((d) => ({
        ...d,
        connectionCheckCompletedAt: new Date(d.connectionCheckCompletedAt),
        connectionCheckStartedAt: new Date(d.connectionCheckStartedAt),
    }));
    const receiverIds = Array.from(
        new Set(sortedData.map((d) => d.receiverId)),
    );

    return receiverIds.map((id) => {
        const entries = sortedData.filter((d) => d.receiverId === id);
        const { organizationName, receiverName } = entries[0];
        const days = eachDayOfInterval(inter).map((day) => {
            const dayEntries = entries.filter(
                (e) =>
                    e.connectionCheckCompletedAt.toLocaleDateString() ===
                    day.toLocaleDateString(),
            );
            const dayString = day.toLocaleDateString();
            const timePeriods = timePeriodLabels.map((time) => {
                const start = new Date(`${day.toLocaleDateString()}, ${time}`);
                const end = addMinutes(start, timePeriodMinutes);

                const timePeriodEntries = dayEntries.filter((e) => {
                    return (
                        e.connectionCheckCompletedAt >= start &&
                        e.connectionCheckCompletedAt < end
                    );
                });
                const agg = timePeriodEntries.reduce(
                    (
                        agg,
                        { connectionCheckSuccessful, connectionCheckResult },
                    ) => {
                        if (connectionCheckSuccessful) agg.success += 1;
                        else {
                            agg.fail += 1;
                            if (
                                connectionCheckResult
                                    .toLowerCase()
                                    .includes(filterResultMessage.toLowerCase())
                            ) {
                                agg.isResultFilterMatch = true;
                            }
                        }
                        return agg;
                    },
                    {
                        success: 0,
                        fail: 0,
                        isResultFilterMatch: false as boolean,
                    },
                );

                const successRateType =
                    agg.success && agg.fail
                        ? SuccessRate.MIXED_SUCCESS
                        : !agg.success && !agg.fail
                          ? SuccessRate.UNDEFINED
                          : agg.success
                            ? SuccessRate.ALL_SUCCESSFUL
                            : SuccessRate.ALL_FAILURE;
                const matchingFilter = !filterResultMessage
                    ? MatchingFilter.NO_FILTER
                    : filterResultMessage && agg.isResultFilterMatch
                      ? MatchingFilter.FILTER_IS_MATCHED
                      : MatchingFilter.FILTER_NOT_MATCHED;

                return {
                    start: start,
                    end: end,
                    ...agg,
                    successRateType,
                    matchingFilter,
                    entries: timePeriodEntries,
                };
            });
            return {
                day,
                dayString,
                timePeriods,
                entries: dayEntries,
            };
        });

        const { success, fail } = days.reduce(
            (agg, curr) => {
                curr.timePeriods.forEach((c) => {
                    agg.fail += c.fail;
                    agg.success += c.success;
                });
                return agg;
            },
            { success: 0, fail: 0 },
        );
        const successRate =
            !!success || !!fail
                ? Math.round((100 * success) / (success + fail))
                : 0;
        const successRateType =
            !success && !fail
                ? SuccessRate.UNDEFINED
                : successRate === 0
                  ? SuccessRate.ALL_FAILURE
                  : successRate === 100
                    ? SuccessRate.ALL_SUCCESSFUL
                    : SuccessRate.MIXED_SUCCESS;

        return {
            organizationName,
            receiverName,
            id,
            successRate,
            successRateType,
            days,
        };
    });
}

export interface ReceiverStatusTimePeriod {
    organizationName: string;
    receiverName: string;
    id: number;
    successRate: number;
    successRateType: SuccessRate;
    days: {
        day: Date;
        dayString: string;
        timePeriods: {
            start: Date;
            end: Date;
            success: number;
            fail: number;
            isResultFilterMatch: boolean;
            successRateType: SuccessRate;
            matchingFilter: MatchingFilter;
            entries: RSReceiverStatusParsed[];
        }[];
        entries: RSReceiverStatusParsed[];
    }[];
}

export type RSReceiverStatusParsed = Omit<
    RSReceiverStatus,
    "connectionCheckStartedAt" | "connectionCheckCompletedAt"
> & {
    connectionCheckStartedAt: Date;
    connectionCheckCompletedAt: Date;
};

export function filterStatuses(
    statuses: ReceiverStatusTimePeriod[],
    receiverNameLike?: string,
    successRateLike?: SuccessRate,
) {
    return statuses.filter(({ receiverName, successRateType }) => {
        const result =
            (!receiverNameLike ||
                receiverName.toLowerCase().includes(receiverNameLike)) &&
            (!successRateLike ||
                successRateLike === SuccessRate.UNDEFINED ||
                successRateType === successRateLike);

        return result;
    });
}
