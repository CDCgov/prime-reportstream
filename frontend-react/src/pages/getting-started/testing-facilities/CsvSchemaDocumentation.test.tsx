import { render } from "../../../utils/CustomRenderUtils";

import {
    CsvSchemaDocumentationItem,
    CsvSchemaItem,
} from "./CsvSchemaDocumentation";

const baseItem: CsvSchemaItem = {
    name: "Sample Item",
    colHeader: "sample_item",
    required: false,
    requested: false,
    acceptedFormat: false,
    acceptedValues: false,
    acceptedExample: false,
    valueType: "",
    values: [],
    notes: [],
};

describe("CsvSchemaDocumentationItem", () => {
    test("render a schema item", () => {
        const { container } = render(
            <CsvSchemaDocumentationItem item={baseItem} />
        );
        expect(container.firstChild).toMatchSnapshot();
    });
});
