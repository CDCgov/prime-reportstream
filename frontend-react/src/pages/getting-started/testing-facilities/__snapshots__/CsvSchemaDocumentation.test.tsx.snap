// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CsvSchemaDocumentationItem render a schema item 1`] = `
<div>
  <h4
    class="font-body-md margin-bottom-2"
    id="doc-sample_item"
  >
    Sample Item
    <span
      class="text-normal bg-white border-1px border-base font-body-3xs padding-x-1 padding-y-05 text-base margin-left-2 text-ttbottom"
    >
      Optional
    </span>
  </h4>
  <div
    class="margin-bottom-3"
  />
  <div
    class="grid-row margin-bottom-05"
  >
    <div
      class="grid-col-4 text-base"
    >
      Column header
    </div>
    <div
      class="grid-col-auto"
    >
      sample_item
    </div>
  </div>
  <div
    class="grid-row margin-bottom-05 border-base-lighter border-top-1px padding-top-1"
  >
    <div
      class="grid-col-4 text-base"
    >
      Value type
    </div>
    <div
      class="grid-col-auto"
    />
  </div>
  <div
    class="grid-row margin-bottom-05 border-base-lighter border-top-1px padding-top-1"
  >
    <div
      class="grid-col-4 text-base"
    />
    <ul
      class="grid-col-8 value-list"
    />
  </div>
</div>
`;
